Libs/MRML/Widgets/qMRMLVolumeThresholdWidget.cxx

```diff
@@ -263,6 +264,10 @@ void qMRMLVolumeThresholdWidget::updateWidgetFromMRMLDisplayNode()
   // even if the values have been just set programmatically/automatically.
   bool wasBlocking = d->VolumeThresholdRangeWidget->blockSignals(true);
 
+  double range[2];
+  d->scalarRange(d->VolumeDisplayNode, range);
+  d->setRange(range[0], range[1]);
```

Comment:
I switched between two volumes with different scalar ranges (MRHead and CTChest) and it worked well. So MRHead was 0 to 297 or something and then CTChest had a lower threshold that was a negative number and it switched to it appropriately.

> The range must be large enough to contain the current threshold value.

This brings up a point I hadn’t considered which is if the scalar range is say 0 to 297 and someone had set the lower threshold to -400 or something, then should the bounds of the threshold widget get set to -400 and 297? I guess so. However should users even be allowed to set a lower threshold below the scalar min value or an upper threshold above the scalar max value? Or should it clamp to this scalar range if they set it to a value outside of the range? Then the threshold range bounds widget would need to support changing bounds only within the scalar range (0 to 297, 50 to 200) and not support cases that go outside the range (-400 to 400).

Reply:
I added comment on exactly this (I hope it did not get lost). The GUI must not ever constrain what you can set in MRML. For example, you must allow setting threshold = 10 for an ultrasound image, and keep that, regardless of what the current image's scalar range is. If you replay an image sequence and happen to come across an all-black image, it would be terrible if it forced the threshold to be 0.

Note that I don't talk about switching between volume nodes, but changing vtkImageData (by sequence replay, updates via OpenIGTLink, etc).

Reply:
Using current Slicer:
```python
import SampleData
volumeNode = SampleData.SampleDataLogic().downloadMRHead()
volumeNode.GetDisplayNode().GetLowerThreshold()
# -32768.0
volumeNode.GetDisplayNode().GetUpperThreshold()
# 32768.0
volumeNode.GetImageData().GetScalarRange()
# (0.0, 279.0)
```
Above shows if I load the MRHead volume into the scene, it has a scalar range of 0 to 297, but the currently set lower threshold and upper threshold is no where near this. The MRML display node has values that are inconsistent with what is shown by the threshold widget. The threshold widget shows -600 to 600 though the actual values by the mrml node is -32768 to 32768. Why does this display volume have an initial lower and upper threshold that is nowhere near the scalar range?

Something would need to change before I change it to the threshold bounds so that they at least include the currently set threshold values as otherwise for the MRHead the widget would show bounds of -32768 to 32768.

Reply:
I guess there is some kind of initialization of the display node when you assign it to the volume node. There the threshold could be initialized to the scalar range. Since thresholding is disabled by default, it should not be a problem that it is tightly set to the scalar range.

Reply:
Ok I'll look into this initialization so then the widget can truly reflect the currently set MRML threshold values.

Reply:
This is where the display node is being initialized to -32768 to 32768 and it not getting updated based on the scalar range when an volume node is connected.
https://github.com/Slicer/Slicer/blob/74d3ec4bb1996e73d82470195847a81703979488/Libs/MRML/Core/vtkMRMLScalarVolumeDisplayNode.cxx#L91

Reply:
@lassoan I've pushed an additional commit to initialize a scalar volume node with a threshold range that matches the scalar range instead of the current VTK_SHORT_MIN and VTK_SHORT_MAX values.